File: src/fairseq2/data/text/sentencepiece.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

from pathlib import Path
from typing import TYPE_CHECKING, List, Optional, Sequence, final

from fairseq2n import DOC_MODE
from torch import Tensor

from fairseq2.assets import default_asset_store, default_download_manager
from fairseq2.data.text.text_tokenizer import (
    StandardTextTokenizerLoader,
    TextTokenDecoder,
    TextTokenEncoder,
    TextTokenizer,
)
from fairseq2.data.vocabulary_info import VocabularyInfo
from fairseq2.typing import Device, override

if TYPE_CHECKING or DOC_MODE:

    @final
    class SentencePieceModel:
        def __init__(
            self, path: Path, control_symbols: Optional[Sequence[str]] = None
        ) -> None:
            ...

        def token_to_index(self, token: str) -> int:
            ...

        def index_to_token(self, idx: int) -> str:
            ...

        @property
        def unk_idx(self) -> Optional[int]:
            ...

        @property
        def bos_idx(self) -> Optional[int]:
            ...

        @property
        def eos_idx(self) -> Optional[int]:
            ...

        @property
        def pad_idx(self) -> Optional[int]:
            ...

        @property
        def vocabulary_size(self) -> int:
            ...

    @final
    class SentencePieceEncoder(TextTokenEncoder):
        def __init__(
            self,
            model: SentencePieceModel,
            prefix_tokens: Optional[Sequence[str]] = None,
            suffix_tokens: Optional[Sequence[str]] = None,
            reverse: bool = False,
            enable_sampling: bool = False,
            nbest_size: int = -1,
            alpha: float = 0.1,
            device: Optional[Device] = None,
            pin_memory: bool = False,
        ) -> None:
            ...

        @override
        def __call__(self, text: str) -> Tensor:
            ...

        @override
        def encode_as_tokens(self, text: str) -> List[str]:
            ...

        @property
        @override
        def prefix_indices(self) -> Optional[Tensor]:
            ...

        @property
        @override
        def suffix_indices(self) -> Optional[Tensor]:
            ...

    @final
    class SentencePieceDecoder(TextTokenDecoder):
        def __init__(self, model: SentencePieceModel, reverse: bool = False) -> None:
            ...

        @override
        def __call__(self, token_indices: Tensor) -> str:
            ...

        @override
        def decode_from_tokens(self, tokens: Sequence[str]) -> str:
            ...

else:
    from fairseq2n.bindings.data.text.sentencepiece import (
        SentencePieceDecoder as SentencePieceDecoder,
    )
    from fairseq2n.bindings.data.text.sentencepiece import (
        SentencePieceEncoder as SentencePieceEncoder,
    )
    from fairseq2n.bindings.data.text.sentencepiece import (
        SentencePieceModel as SentencePieceModel,
    )

    # Ensure that extension types are virtual subclasses of their corresponding
    # abstract base types.
    TextTokenEncoder.register(SentencePieceEncoder)
    TextTokenDecoder.register(SentencePieceDecoder)

    def _set_module_name() -> None:
        for t in [SentencePieceDecoder, SentencePieceEncoder, SentencePieceModel]:
            t.__module__ = __name__

    _set_module_name()


class SentencePieceTokenizer(TextTokenizer):
    """Represents a SentencePiece tokenizer."""

    _model: SentencePieceModel

    def __init__(
        self, path: Path, control_symbols: Optional[Sequence[str]] = None
    ) -> None:
        """
        :param path:
            The path to the SentencePiece model file.
        :param control_symbols:
            The list of control symbols to add to the SentencePiece model.
        """
        self._model = SentencePieceModel(path, control_symbols)

        vocab_info = vocab_info_from_sentencepiece(self._model)

        super().__init__(vocab_info)

    @override
    def create_raw_encoder(
        self, *, device: Optional[Device] = None, pin_memory: bool = False
    ) -> SentencePieceEncoder:
        return SentencePieceEncoder(self._model, device=device, pin_memory=pin_memory)

    @override
    def create_decoder(self) -> SentencePieceDecoder:
        return SentencePieceDecoder(self._model)

    @final
    @property
    def model(self) -> SentencePieceModel:
        return self._model


@final
class BasicSentencePieceTokenizer(SentencePieceTokenizer):
    """Represents a SentencePiece tokenizer that encodes text with BOS and EOS."""

    def __init__(self, path: Path) -> None:
        """
        :param path:
            The path to the SentencePiece model file.
        """
        super().__init__(path)

    @override
    def create_encoder(
        self,
        *,
        task: Optional[str] = None,
        lang: Optional[str] = None,
        mode: Optional[str] = None,
        device: Optional[Device] = None,
        pin_memory: bool = False,
    ) -> SentencePieceEncoder:
        """Create a token encoder.

        :param task:
            Not used.
        :param lang:
            Not used.
        :param mode:
            Must be 'default' or 'prompt'. If ``None``, defaults to 'default'.
        :param device:
            The device on which to construct tensors.
        :param pin_memory:
            If ``True``, uses pinned memory while constructing tensors.
        """
        if task is not None:
            raise ValueError(f"`task` must be `None`, but is '{task}' instead.")

        if lang is not None:
            raise ValueError(f"`lang` must be `None`, but is '{lang}' instead.")

        if mode is None or mode == "default":
            prefix_tokens = ["<s>"]
            suffix_tokens = ["</s>"]
        elif mode == "prompt":
            prefix_tokens = ["<s>"]
            # In prompt mode, we expect the generator to finish the sequence.
            suffix_tokens = None
        else:
            raise ValueError(
                f"`mode` must be 'default' or 'prompt', but is '{mode}' instead."
            )

        return SentencePieceEncoder(
            self._model,
            prefix_tokens=prefix_tokens,
            suffix_tokens=suffix_tokens,
            device=device,
            pin_memory=pin_memory,
        )


load_basic_sentencepiece_tokenizer = StandardTextTokenizerLoader(
    default_asset_store,
    default_download_manager,
    lambda path, _: BasicSentencePieceTokenizer(path),
)


def vocab_info_from_sentencepiece(model: SentencePieceModel) -> VocabularyInfo:
    """Return the vocabulary information of ``model``."""
    return VocabularyInfo(
        model.vocabulary_size,
        model.unk_idx,
        model.bos_idx,
        model.eos_idx,
        model.pad_idx,
    )

File: native/python/src/fairseq2n/bindings/data/text/sentencepiece.cc
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

#include "fairseq2n/bindings/module.h"

#include <cstdint>
#include <memory>
#include <optional>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

#include <fairseq2n/data/text/sentencepiece/sentencepiece.h>

namespace py = pybind11;

namespace fairseq2n {

void
def_sentencepiece(py::module_ &text_module)
{
    py::module_ m = text_module.def_submodule("sentencepiece");

    py::class_<sp_model, std::shared_ptr<sp_model>>(m, "SentencePieceModel")
        .def(
            py::init([](
                std::string_view pathname,
                std::optional<std::vector<std::string>> maybe_control_symbols)
            {
                sp_model_options opts{};

                if (maybe_control_symbols)
                    opts.control_symbols() = *std::move(maybe_control_symbols);

                return std::make_shared<sp_model>(pathname, std::move(opts));
            }),
            py::arg("pathname"),
            py::arg("control_symbols") = std::nullopt)

        .def(
            py::pickle(
                [](const sp_model &self)
                {
                    std::string serialized = self.serialize();

                    return py::bytes(serialized);
                },
                [](const py::bytes &bytes)
                {
                    auto serialized = bytes.cast<std::string>();

                    return sp_model::from_serialized(serialized);
                }))

        .def("token_to_index", &sp_model::token_to_index)
        .def("index_to_token", &sp_model::index_to_token)

        .def_property_readonly("unk_idx", &sp_model::unk_idx)
        .def_property_readonly("bos_idx", &sp_model::bos_idx)
        .def_property_readonly("eos_idx", &sp_model::eos_idx)
        .def_property_readonly("pad_idx", &sp_model::pad_idx)

        .def_property_readonly("vocabulary_size", &sp_model::vocabulary_size);

    py::class_<sp_encoder, std::shared_ptr<sp_encoder>>(m, "SentencePieceEncoder")
        .def(
            py::init([](
                std::shared_ptr<const sp_model> model,
                std::optional<std::vector<std::string>> maybe_prefix_tokens,
                std::optional<std::vector<std::string>> maybe_suffix_tokens,
                bool reverse,
                bool enable_sampling,
                std::int32_t nbest_size,
                float alpha,
                std::optional<at::Device> maybe_device,
                bool pin_memory)
            {
                auto opts = sp_encoder_options()
                    .reverse(reverse)
                    .enable_sampling(enable_sampling)
                    .nbest_size(nbest_size)
                    .alpha(alpha)
                    .maybe_device(maybe_device)
                    .pin_memory(pin_memory);

                if (maybe_prefix_tokens)
                    opts.prefix_tokens() = *std::move(maybe_prefix_tokens);

                if (maybe_suffix_tokens)
                    opts.suffix_tokens() = *std::move(maybe_suffix_tokens);

                return std::make_shared<sp_encoder>(std::move(model), std::move(opts));
            }),
            py::arg("model"),
            py::arg("prefix_tokens")   = std::nullopt,
            py::arg("suffix_tokens")   = std::nullopt,
            py::arg("reverse")         = false,
            py::arg("enable_sampling") = false,
            py::arg("nbest_size")      = -1,
            py::arg("alpha")           = 0.1,
            py::arg("device")          = std::nullopt,
            py::arg("pin_memory")      = false)
        .def(
            "__call__",
            &sp_encoder::operator(),
            py::arg("text"),
            py::call_guard<py::gil_scoped_release>{})
        .def(
            "encode_as_tokens",
            &sp_encoder::encode_as_tokens,
            py::arg("text"),
            py::call_guard<py::gil_scoped_release>{})

        .def_property_readonly("prefix_indices", &sp_encoder::prefix_indices)
        .def_property_readonly("suffix_indices", &sp_encoder::suffix_indices);

    py::class_<sp_decoder, std::shared_ptr<sp_decoder>>(m, "SentencePieceDecoder")
        .def(
            py::init<std::shared_ptr<const sp_model>, bool>(),
            py::arg("model"),
            py::arg("reverse") = false)
        .def(
            "__call__",
            &sp_decoder::operator(),
            py::arg("token_indices"),
            py::call_guard<py::gil_scoped_release>{})
        .def(
            "decode_from_tokens",
            &sp_decoder::decode_from_tokens,
            py::arg("tokens"),
            py::call_guard<py::gil_scoped_release>{});

    map_functors().register_<sp_encoder>();
    map_functors().register_<sp_decoder>();
}

}  // namespace fairseq2n
